# Lập trình cho Khoa học dữ liệu: Đồ án cuối kì
## Nhóm 8: Thông tin sinh viên
- #### Trello: https://trello.com/b/Tb6Sjg8O


| MSSV     | Họ tên            | GitHub                         |    Email         
|--------------|---------------------|------------------------------ |--------------------------------------------------|
| 19127009 | Lữ Thế Vỹ | [luthevy](https://github.com/luthevy) | 19127009@student.hcmus.edu.vn
| 19127416 | Mạc Văn Hưng     | [machung2001](https://github.com/machung2001)        |19127416@student.hcmus.edu.vn


# Mục lục
- #### [Lời nói đầu](#Thu-thập-dữ-liệu)
    - [Giới thiệu](#Thu-thập-dữ-liệu)
    - [Mục tiêu](#Thu-thập-dữ-liệu)
    - [Nguồn tài liệu](#Thu-thập-dữ-liệu)
- #### [Bước chuẩn bị](#Thu-thập-dữ-liệu)
    - [Nhận diện môi trường code](#Thu-thập-dữ-liệu)
    - [Import thư viện](#Thu-thập-dữ-liệu)
- #### [Khám phá dữ liệu](#Thu-thập-dữ-liệu)
    - [Đọc dữ liệu](#Thu-thập-dữ-liệu)
    - [Thông tin cơ bản](#Thu-thập-dữ-liệu)
        - [Số dòng, số cột của tập dữ liệu](#Thu-thập-dữ-liệu)
        - [Ý nghĩa từng cột](#Thu-thập-dữ-liệu)
        - [Kiểu dữ liệu từng cột](#Thu-thập-dữ-liệu)
    - [Nhận xét](#Thu-thập-dữ-liệu)
        - [Có cột nào cần loại bỏ không?](#Thu-thập-dữ-liệu)
        - [Có dòng dữ liệu nào cần loại bỏ không?](#Thu-thập-dữ-liệu)
        - [Có cột nào bất thường không?](#Thu-thập-dữ-liệu)
        - [Có dòng dữ liệu nào bất thường không?](#Thu-thập-dữ-liệu)
        - [Có dòng dữ liệu nào bị lặp không?](#Thu-thập-dữ-liệu)
        - [Có dòng dữ liệu nào mang ý nghĩa khác không?](#Thu-thập-dữ-liệu)
- #### [Khám phá dữ liệu (tiếp theo) - Đặt câu hỏi](#Thu-thập-dữ-liệu)
    - [Phân bố dữ liệu kiểu numerical](#Thu-thập-dữ-liệu)
    - [Phân bố dữ liệu kiểu categorical](#Thu-thập-dữ-liệu)



## Lời nói đầu

## Môi trường code

In [2]:
import sys
sys.executable

'/Users/lanmac/opt/anaconda3/envs/min_ds-env/bin/python'

## Import các thư viện cần thiết

In [3]:
# Import thư viện cần dùng
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

---

## Khám phá dữ liệu

---

### Đọc dữ liệu từ file csv vào dataframe

In [5]:
hcs_df = pd.read_csv('data/healthcare-dataset-stroke-data.csv')
hcs_df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [9]:
num_rows = hcs_df.shape[0]
num_cols = hcs_df.shape[1]

hcs_df.shape

(5110, 12)

Dữ liệu có 12 cột và 5110 dòng.

### Mỗi cột có ý nghĩa gì?

In [33]:
with open('data/description.txt', 'r') as f:
    print(f.read())

1) id:                 unique identifier
2) gender:             "Male", "Female" or "Other"
3) age:                age of the patient
4) hypertension:       0 if the patient doesn't have hypertension, 1 if the patient has hypertension
5) heart_disease:      0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
6) ever_married:       "No" or "Yes"
7) work_type:          "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
8) Residence_type:     "Rural" or "Urban"
9) avg_glucose_level:  average glucose level in blood
10) bmi:               body mass index
11) smoking_status:    "formerly smoked", "never smoked", "smokes" or "Unknown"*
12) stroke.            1 if the patient had a stroke or 0 if not
*Note: "Unknown" in smoking_status means that the information is unavailable for this patient


- **id**: mã id
- **gender**: giới tính ("Male", Female" và "Other")
- **age**: tuổi tác
- **hypertension**: bị huyết áp cao (0 là không, 1 là có)
- **heart_disease**: bị bệnh tim (0 là không, 1 là có)
- **ever_married**: tình trạng kết hôn ("No" và "Yes")
- **work_type**: loại nghề nghiệp ("children", "Govt_jov", "Never_worked", "Private" và "Self-employed")
- **Residence_type**: loại cư dân ("Rural" hoặc "Urban")
- **avg_glucose_level**: lượng đường trung bình trong máu
- **bmi**: chỉ số khối cơ thể
- **smoking_status**: tình trạng hút thuốc ("formerly smoked", "never smoked", "smokes" và "Unknown")
- **stroke**: bị đột quỵ (0 là không, 1 là có)

### Mỗi cột hiện đang có kiểu dữ liệu gì?

In [13]:
hcs_df.dtypes

id                     int64
gender                object
age                  float64
hypertension           int64
heart_disease          int64
ever_married          object
work_type             object
Residence_type        object
avg_glucose_level    float64
bmi                  float64
smoking_status        object
stroke                 int64
dtype: object

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

Hiện có 6 cột thuộc dạng numeric: "age", "hypertension", "heart_disease", "avg_glucose_level", "bmi", "stroke". Nhưng 3 cột "hypertension", "heart_disease", "stroke" do mã được đánh dưới dạng số nên 3 cột này sẽ là dạng categorical. Vì vậy, chỉ có 3 cột numeric là "age", "avg_glucose_level", "bmi". 

In [27]:
def missing_ratio(col):
    return col.isna().mean() * 100

In [34]:
num_col_profiles_df = hcs_df[["age", "avg_glucose_level", "bmi"]]
num_col_profiles_df = num_col_profiles_df.agg([missing_ratio, pd.Series.min, pd.Series.max])
num_col_profiles_df

age  avg_glucose_level        bmi
missing_ratio   0.00               0.00   3.933464
min             0.08              55.12  10.300000
max            82.00             271.74  97.600000

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

Ban đầu sẽ có 5 cột thuộc dạng categorical: "gender", "ever_married", "work_type", "Residence_type", "smoking_status".
Bởi vì lý do ở trên nên sẽ có thêm 3 cột: "hypertension", "heart_disease", "stroke". Vì vậy, tổng cộng có 8 cột thuộc dạng categorical: "gender", "hypertension", "heart_disease", "ever_married", "work_type", "Residence_type", "smoking_status", "stroke".

In [29]:
def unique_non_null(s):
    return s.dropna().unique()

In [35]:
cate_col_profiles_df = hcs_df[["gender", "hypertension", "heart_disease","work_type", 
                               "Residence_type", "smoking_status", "stroke"]]
cate_col_profiles_df = cate_col_profiles_df.agg([missing_ratio, unique_non_null, pd.Series.nunique])
cate_col_profiles_df = cate_col_profiles_df.rename(index={'unique_non_null': 'diff_vals', 'nunique': 'num_diff_vals'})

## Phân tích dữ liệu để trả lời cho các câu hỏi 

### Giữa những người đã từng hoặc đang hút thuốc và những người chưa hút thuốc bao giờ thì tỷ lệ bị mắc bệnh chênh lệch bao nhiêu?

- Để lấy thông tin người này có bị mắc bệnh không, ta xét 3 cột hypertension, heart_disease, stroke, nếu 1 trong 3 cột nếu có bị (tương đương số 1) thì ta cho người này có bệnh (True).
- Từ đó ta tạo thêm cột has_disease để lưu boolean (True, False) rằng người này có bệnh hay không

In [448]:
copy_hcs_df = hcs_df.copy()
has_disease = (hcs_df['hypertension'] == 1) | (hcs_df['heart_disease'] == 1) | (hcs_df['stroke'] == 1)
copy_hcs_df['has_disease'] = has_disease

copy_hcs_df['has_disease']

0        True
1        True
2        True
3        True
4        True
        ...  
5105     True
5106    False
5107    False
5108    False
5109    False
Name: has_disease, Length: 5110, dtype: bool

- Ta group 2 cột smoking_status và has_disease vừa được tạo ở bước trên để tính phân bố dựa theo smoking_status
- Sau đó tính tỷ lệ mắc bệnh bằng cách chia tổng của từng loại smoking_status

In [446]:
disease_smoke_df = copy_hcs_df.groupby(['smoking_status','has_disease']).size()
disease_smoke_df = disease_smoke_df.unstack(0)
disease_smoke_df = disease_smoke_df / disease_smoke_df.sum()
disease_smoke_df

smoking_status   Unknown  formerly smoked  never smoked    smokes
has_disease                                                      
False           0.915803         0.752542      0.815539  0.795944
True            0.084197         0.247458      0.184461  0.204056

Theo số liệu được tính toán như trên, ta thấy tỷ lệ mắc bệnh (hypertension hoặc heart disease hoặc stroke) của những người không hút thuốc (never smoked) thấp hơn so với những người đã từng (formerly smoked) hoặc đang hút thuốc (smokes). Ta còn thấy tỷ lệ không mắc bệnh của never smoked lại cao hơn 2 loại kia. Từ những điều trên cho thấy hành vi hút thuốc có tác động rất lớn đến sức khoẻ của con người và tỷ lệ mắc bệnh cao lên.

### Con người càng già thì càng dễ bị mắc bệnh không?